### PR TITLE
Allow stan protocol to set native *stan.Msg as a context value

### PR DESCRIPTION
**Pull request by dan-j**
See #657 

This is an approach which doesn't introduce any breaking changes, if users do nothing their `.Receive()` functions will receive the same `ctx` object they always have, however using the new `PropogateContext()` option will override it.

I'm going to add some comments to other parts of the codebase in this MR because I think I have an alternative approach, but needs discussion

Signed-off-by: dan-j <5727701+dan-j@users.noreply.github.com>

### CODE REVIEW

File: v2/client/invoker.go [hunk 1]
```diff
@@ -118,12 +118,18 @@ func (r *receiveInvoker) IsResponder() bool {
...
-		return mctx.Context()
-	} else {
+		ctx = mctx.Context()
+	}
+
+	if ctx == nil {
 		return fallback
 	}
+
+	return ctx
```
Comment (dan-j) [inline v2/client/invoker.go:134]: changes to this function means that if `MessageContext.Context()` returns `nil`, then `fallback` is returned, whereas previously it would return `nil`. 

Not sure if we can do this as it may be a breaking change, however I'd say this is safer since using a `nil` context is typically a bad idea anyway.

**Comment by dan-j** [comment]
An alternative approach, but not sure of the consequences:

Is there a reason why `context.Background()` is passed in here? If we used `ctx` instead, then we can get rid of the `PropogateContext()` option.

https://github.com/cloudevents/sdk-go/blob/93f9fa1cd3103d71d1b609a75ec3160ed601b436/v2/binding/to_event.go#L45-L52

**Comment by slinkydeveloper** [comment]
> Is there a reason why context.Background() is passed in here? If we used ctx instead, then we can get rid of the PropogateContext() option.

AFAIR I pass `context.Background()` here to avoid propagating any option related to the writes (ForceBinary, ForceStructured) which will prevent this DirectWrite to work

File: v2/client/invoker.go (hunk 1 above)
Comment (slinkydeveloper) [reply v2/client/invoker.go:134]: I'm ok with this change, thus as you outlined before, using another context is a breaking change

**Review by slinkydeveloper — COMMENTED** [review]
I'm ok with this change, although before merging it, I would love to find out if we can avoid swapping context, but more "decorating it"

**Comment by slinkydeveloper** [comment]
@dan-j Look at https://github.com/cloudevents/sdk-go/pull/659, the idea of this PR is to reuse the same context, jut wrap in it the additional informations we want to propagate from the message up to the client

**Comment by slinkydeveloper** [comment]
I think we should not propagate the whole stan messages, but only the metadata we're interested in. How does it sounds?

**Review by n3wscott — APPROVED** [review]

**Comment by dan-j** [comment]
> I think we should not propagate the whole stan messages, but only the metadata we're interested in. How does it sounds?

Sure, the three things which come to mind are: `Redelivered`, `RedeliveryCount` and `Sequence`.

Do you think these should be standard context options and the getters/setters somewhere in the core SDK, or keep them contained within the `stan` package? I know protocols such as http won't be able to support this functionality but others may, and since they're values on context consumers can't expect they be set, but can check if they are and act appropriately.

**Comment by slinkydeveloper** [comment]
> keep them contained within the stan package?

Let's keep them in the stan package

**Comment by dan-j** [comment]
@slinkydeveloper 

OK, I've updated this PR to expose a new context decorator which can be used with your new `WithInboundContextDecorator()` option. This still sets the whole message on the context, but uses an unexported key so the data is only available by using the accessor functions in `context.go`.

Thoughts?

File: protocol/stan/v2/context.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			// we store the raw *stan.Msg on the context, but under a unexported key so only the below accessor
+			// functions can be used to extract this info.
+			return context.WithValue(ctx, msgKey, msg.Msg)
+		}
+
+		return ctx
+	}
+}
+
+func SequenceFrom(ctx context.Context) (uint64, bool) {
```
Comment (slinkydeveloper) [inline protocol/stan/v2/context.go:27]: Can you add a godoc to this and below methods?

**Review by slinkydeveloper — COMMENTED** [review]
That's good! Can you add the godocs to the methods?

File: protocol/stan/v2/context.go (hunk 2 above)
Comment (dan-j) [reply protocol/stan/v2/context.go]: done, I've actually changed the behaviour to keep all the data in a single struct, seems unnecessary to have multiple functions to extract this data because you would have ended up with multiple `if value, ok := FooFrom(ctx); ok { ... }` statements if you wanted to view all of the properties.

**Review by slinkydeveloper — COMMENTED** [review]
One last thing, can you add a small fancy test for it? Just add it to the actual test an "assert" that MsgMetadata is not empty

**Comment by dan-j** [comment]
Added tests with a valid `*stan.Message` and one where it is not a `*stan.Message`

**Review by slinkydeveloper — APPROVED** [review]
